### ISSUE TITLE
Sound is very low and missing mid/bass frequencies

### ISSUE THREAD

**Issue by karypid** [karypid/YogaPro-16IMH9#2]
The audio even at maximum volume is very weak. The sound seems to lack bass and mid frequencies.

**Comment by karypid** [comment]
This is due to https://bugzilla.kernel.org/show_bug.cgi?id=217449

You can use the script from https://bugzilla.kernel.org/attachment.cgi?id=304763 to fix the speakers.

I use this systemd unit that runs on startup as a workaround:

```
❯ cat /etc/systemd/system/yoga-16imh9-speakers.service 
[Unit]
Description=Turn on speakers using i2c configuration

[Service]
User=root
Type=oneshot
ExecStart=/bin/sh -c "/var/personalization/sbin/2pa-byps.sh 2 | logger"

[Install]
WantedBy=multi-user.target
```

**Comment by rtoscanelli** [comment]
The sound with the script provided works, however is not as good comparing it with the machine running on Windows.

**Comment by tedonporter** [comment]
> The sound with the script provided works, however is not as good comparing it with the machine running on Windows.

The following repo provides some presets for EasyEffects that corrects this and makes the audio sound wayyy better. Pretty close to the experience I was getting on windows. I would recommend downloading EasyEffects via flathub as it comes pre-bundled with the EQ plugins needed for the presets. Also, I would set the option in the app settings to have it always start on startup.

https://github.com/maximmaxim345/yoga_pro_9i_gen9_linux

**Comment by karypid** [comment]
> The following repo provides some presets for EasyEffects that corrects this and makes the audio sound wayyy better. Pretty close to the experience I was getting on windows. I would recommend downloading EasyEffects via flathub as it comes pre-bundled with the EQ plugins needed for the presets. Also, I would set the option in the app settings to have it always start on startup.
> 
> https://github.com/maximmaxim345/yoga_pro_9i_gen9_linux

Wow, thank you for this. I just ran down the rabbit hole of [the tuning guide](https://wwmm.github.io/easyeffects/guide_1.html) and created a profile that is more to my liking!

Here it is: [YogaPro9i-2024.json](https://github.com/user-attachments/files/16116290/YogaPro9i-2024.json)

I think most people may find this a bit heavy on the high frequencies (which I like) but it's definitely improved my experience. Thank you for sharing!

**Comment by alyraffauf** [comment]
Sorry to revive this -- but has anyone else experienced the audio disappearing after some time? The script linked fixes my audio, and it sounds great, but two things happen:

1. Audio simply disappears after a while, and returns to a tweeter-only state. I think this is an issue of the speakers being put to sleep, and I might have fixed it with udev rules.
2. Under heavy load, for example, when launching a game, I'm losing audio entirely. Doesn't come back until I quit the game (and then sometimes with the subwoofers suspended)

I'm on kernel 6.12, looks like people on the upstream bugzilla issue are reporting similar problems..

**Comment by karypid** [comment]
> I'm on kernel 6.12, looks like people on the upstream bugzilla issue are reporting similar problems..

Yes, you are correct that there are additional issues in newest 6.12 kernels, but I see your comment 15 with further suggested [workaround to disable power saving](https://bugzilla.kernel.org/show_bug.cgi?id=217449#c15)

I just started using this today after upgrading to a 6.12 kernel and it seems ~to work~ to partially improve the situation, but I still get the speaker output to disappear after some suspend/resume cycles.

```
❯ cat /etc/modprobe.d/yoga-disable-hda-power-save.conf 
options snd_hda_intel power_save=0


~ 
❯ cat /etc/udev/rules.d/99-yoga-i2c-power-control.rules 
SUBSYSTEM=="i2c-dev", ATTR{name}=="Synopsys DesignWare I2C adapter", ATTR{power/control}="on"
```

Can you confirm this has been working well for you for the past month?

**Comment by alyraffauf** [comment]
Hi -- sorry for the late response.

I came up with a much more durable solution [here](https://github.com/maximmaxim345/yoga_pro_9i_gen9_linux/issues/10#issuecomment-2603079359), and found a much more accurate bug report coming this issue. Quoting it bellow, with some links.

> Hey there,
> 
> I've actually come up with a much more durable solution and explanation of the issue.
> 
> Wither kenel 6.12.x, your speakers should work out of the box. But you will quickly find out that they stop working almost immediately after boot, even when using the i2c script.
> 
> This because the kernel TAS2781 amp (which powers your speakers) driver has, from what I can tell, completely broken runtime power management. While the i2c script worked fine alone under 6.11, it's now conflicting with `snd_hda_scodec_tas2781_i2c`. You can test this by reloading that module with `sudo modprobe -r snd_hda_scodec_tas2781_i2c && sudo modprobe snd_hda_scodec_tas2781_i2c`. Your speakers should work, but as soon as they suspend themselves you'll be back to tweeter-only.
> 
> What you need to do is to blacklist the `snd_hda_scodec_tas2781_i2c` module in your modprobe config. You can add a file to `/etc/modprobe.d`with the line `blacklist snd_hda_scodec_tas2781_i2c`, then run the script given in this repository as normal and it should persist across supend/resume cycles, as long as you run it every boot.
> 
> It's necessary to blacklist this module because you will have absolutely erratic speaker behavior otherwise. The kernel module conflcits with our i2c script and seems genuinely quite unstable.
> 
> I've documented this [here](https://github.com/alyraffauf/nixcfg/tree/master/modules/hardware/lenovo/yoga/16IMH9), though my configuration is for NixOS.
> 
> I'd encourage you to read and comment on [this](https://bugzilla.kernel.org/show_bug.cgi?id=219452) kernel bug report in hopes of getting a fix. I've tried writing some kernel patches, but I'm not kernel developer, thus haven't had much luck. This seems to work great for now.
